randovania/resolver/resolver_reach.py

```diff
@@ -143,12 +144,24 @@ def calculate_reach(cls, logic: Logic, initial_state: State) -> ResolverReach:
                     )
                     path_to_node[target_node_index] = list(path_to_node[node_index])
                     path_to_node[target_node_index].append(node_index)
+                    satisfied_requirement_on_node[target_node_index] = (
+                        requirement_including_leaving
+                        if node.heal
+                        else RequirementAnd(
+                            [requirement_including_leaving, satisfied_requirement_on_node[node.node_index]]
+                        )
+                    )
 
                 elif target_node:
                     # If we can't go to this node, store the reason in order to build the satisfiable requirements.
                     # Note we ignore the 'additional requirements' here because it'll be added on the end.
                     if not requirement.satisfied(context, damage_health):
-                        requirements_excluding_leaving_by_node[target_node_index].append(requirement)
+                        full_requirement_for_target = (
+                            requirement
+                            if node.heal
+                            else RequirementAnd([requirement, satisfied_requirement_on_node[node.node_index]])
+                        )
```

Comment:
since you've repeated this twice, maybe break it out into a helper function

Reply:
I moved it to a helper function, and then I later realized that duplicating this is an error. Keeping the helper function seems like a good idea though.